pkg/preflight/aws.go

```diff
@@ -64,90 +71,90 @@ func (a *aws) CheckCloudDrivePermission(cluster *corev1.StorageCluster) error {
-		if !strings.Contains(err.Error(), errAlreadyDetached) {
-			logrus.WithError(err).Errorf("preflight failed to detach volume")
-			return err
-		}
+	// Expand volume
+	if _, err := a.ops.Expand(*vol.VolumeId, uint64(2), dryRunOption); !dryRunSucceeded(err) {
+		return fmt.Errorf("preflight volume expansion dry run failed: %v", err)
 	}
 
-	if err := a.ops.Delete(*vol.VolumeId); err != nil {
-		logrus.WithError(err).Errorf("preflight failed to delete volume")
-		return err
+	// Apply and remove tags
+	tags := map[string]string{
+		"foo": "bar",
+	}
+	if err := a.ops.ApplyTags(*vol.VolumeId, tags, dryRunOption); !dryRunSucceeded(err) {
+		return fmt.Errorf("preflight volume tag dry run failed: %v", err)
+	}
+	if err := a.ops.RemoveTags(*vol.VolumeId, tags, dryRunOption); !dryRunSucceeded(err) {
+		return fmt.Errorf("preflight volume remove tag dry run failed: %v", err)
+	}
+
+	// Detach volume
+	if err := a.ops.Detach(*vol.VolumeId, dryRunOption); !dryRunSucceeded(err) {
+		return fmt.Errorf("preflight volume detach dry run failed: %v", err)
+	}
+
+	// Delete volume
+	// Check permission first then do the actual deletion in cleanup phase
+	if err := a.ops.Delete(*vol.VolumeId, dryRunOption); !dryRunSucceeded(err) {
+		return fmt.Errorf("preflight volume delete dry run failed: %v", err)
 	}
 
-	// Do a cleanup when preflight passed, as it's guaranteed to have delete permission
-	a.cleanupPreflightVolumes(cluster.Name)
+	// Do a cleanup when preflight passed, as it's guaranteed to have permissions for volume deletion
+	// Will delete volumes created in previous attempts as well by cluster ID label
+	a.cleanupPreflightVolumes(volumes)
 	logrus.Infof("preflight check for eks cloud permission passed")
 	return nil
 }
 
-func (a *aws) cleanupPreflightVolumes(clusterName string) {
-	logrus.Infof("preflight cleaning up volumes created in permission check")
-	result, err := a.ops.Enumerate(nil, map[string]string{labelClusterID: clusterName}, "")
-	if err != nil {
-		logrus.Warnf("preflight failed to list volumes to clean up")
-		return
-	}
-
-	logrus.Infof("preflight found %v volumes to clean up", len(result[cloudops.SetIdentifierNone]))
-	for _, v := range result[cloudops.SetIdentifierNone] {
+func (a *aws) cleanupPreflightVolumes(volumes []interface{}) {
+	logrus.Infof("preflight cleaning up volumes created in permission check, %v volumes to delete", len(volumes))
+	for _, v := range volumes {
 		vol := *v.(*ec2.Volume).VolumeId
-		err := a.ops.Detach(vol)
-		if err == nil {
-			logrus.Infof("preflight volume %s detached", vol)
-		}
-		err = a.ops.Delete(vol)
-		if err == nil {
-			logrus.Infof("preflight volume %s deleted", vol)
+		err := a.ops.Delete(vol, nil)
+		if err != nil {
+			logrus.Warnf("preflight failed to delete volume %s: %v", vol, err)
 		}
 	}
 }
+
+func dryRunSucceeded(err error) bool {
+	return err != nil && strings.Contains(err.Error(), dryRunErrMsg)
```

Comment:
The AWS API returns an error even on success?

Reply:
yes, the error type will be `DryRunOperation`